v3/src/components/graph/components/dotplotdots.tsx

```diff
@@ -211,41 +215,84 @@ export const DotPlotDots = observer(function DotPlotDots(props: PlotProps) {
           }
         }
 
-      const
-        // Note that we can get null for either or both of the next two functions. It just means that we have
-        // a circle for the case, but we're not plotting it.
-        getPrimaryScreenCoord = (anID: string) => {
-          const primaryCoord = primaryAxisScale(dataset?.getNumeric(anID, primaryAttrID) ?? -1) /
-              numExtraPrimaryBands,
-            extraPrimaryValue = dataset?.getStrValue(anID, extraPrimaryAttrID),
-            extraPrimaryCoord = extraPrimaryValue
-              ? extraPrimaryAxisScale(extraPrimaryValue ?? '__main__') ?? 0
-              : 0
-          return primaryCoord + extraPrimaryCoord
-        },
-        getSecondaryScreenCoord = (anID: string) => {
-          const secondaryCat = binMap[anID].category,
-            extraSecondaryCat = binMap[anID].extraCategory,
-            indexInBin = binMap[anID].indexInBin,
-            onePixelOffset = primaryIsBottom ? -1 : 1 // Separate circles from axis line by 1 pixel
-          return binMap[anID]
-            ? computeSecondaryCoord({secondaryCat, extraSecondaryCat, indexInBin}) + onePixelOffset
-            : null
-        },
-        getScreenX = primaryIsBottom ? getPrimaryScreenCoord : getSecondaryScreenCoord,
-        getScreenY = primaryIsBottom ? getSecondaryScreenCoord : getPrimaryScreenCoord,
-        getLegendColor = dataConfiguration?.attributeID('legend')
-          ? dataConfiguration?.getLegendColorForCase : undefined
+      const computePrimaryCoord = (anID: string) => {
+        const primaryCoord = primaryAxisScale(dataset?.getNumeric(anID, primaryAttrID) ?? -1) / numExtraPrimaryBands
+        const extraPrimaryValue = dataset?.getStrValue(anID, extraPrimaryAttrID)
+        const extraPrimaryCoord = extraPrimaryValue ? extraPrimaryAxisScale(extraPrimaryValue ?? '__main__') ?? 0 : 0
+        return { primaryCoord, extraPrimaryCoord }
+      }
+
+      const getSubPlotDetails = (anID: string) => {
+        const subPlotKey = dataset?.subPlotKey(anID, secondaryAttrID, extraSecondaryAttrID, extraPrimaryAttrID) ?? {}
+        const casesInCategory = dataConfig?.subPlotCases(subPlotKey) ?? []
+        const caseIndex = casesInCategory.findIndex((id) => id === anID)
+        return { subPlotKey, casesInCategory, caseIndex }
+      }
```

Comment:
The id-to-index lookup here is O(2n^2), since `getSubPlotDetails()` is called twice for each case, which could be a potential performance issue.  I played around a bit and the following caches the results on the first call which should improve performance, particularly on large data sets.
```suggestion
      interface ISubPlotDetails {
        cases: string[]
        indices: Record<string, number>
      }
      const subPlotDetails = new Map<string, ISubPlotDetails>()

      const getSubPlotDetails = (anID: string) => {
        const subPlotKey = dataset?.subPlotKey(anID, secondaryAttrID, extraSecondaryAttrID, extraPrimaryAttrID) ?? {}
        const subPlotMapKey = JSON.stringify(subPlotKey)
        let details: ISubPlotDetails | undefined = subPlotDetails.get(subPlotMapKey)
        if (!details) {
          const cases = dataConfig?.subPlotCases(subPlotKey) ?? []
          const indices: Record<string, number> = {}
          cases.forEach((caseId, index) => indices[caseId] = index)
          details = { cases, indices }
          subPlotDetails.set(subPlotMapKey, details)
        }
        return { subPlotKey, casesInCategory: details.cases, caseIndex: details.indices[anID] }
      }
```
This suggestion doesn't include a related point made elsewhere, that the `subPlotKey` method should be on the `GraphDataConfiguration` rather than on the `DataSet`.

Reply:
This works well -- thanks!